### PR TITLE
fix: throw RangeError on BigInt overflow past ±2^1023 instead of silently wrapping (#6073)

### LINKED ISSUE #6073: BigInt is a fixed 1024-bit ring — silent wraparound past ±2^1023
**Severity:** P1 (silent numeric corruption)
**Found by:** audit `fable-audit-perry-2.md`, confirmed by probe.

### What
BigInt is a fixed 1024-bit two's-complement value; arithmetic past ±2¹⁰²³ silently wraps like an `i1024` instead of being arbitrary precision.

```js
console.log((2n ** 1024n) > 1e308);   // node: true   perry: false  (2^1024 overflowed the sign bit)
```
Any product/power exceeding 2¹⁰²³ (e.g. RSA/2048-bit-curve arithmetic) corrupts with no throw.

### Why
`crates/perry-runtime/src/bigint.rs:10-13`:
```rust
pub const BIGINT_LIMBS: usize = 16; // 1024 bits keeps the sign bit at bit 1023
```

### Fix
Detect overflow and escalate to heap arbitrary precision (e.g. `num-bigint`), or — as an interim correctness stopgap — throw `RangeError` on overflow instead of silently wrapping. Small-BigInt fast path can stay.

**Confidence:** high (probe + source).

### PR DESCRIPTION
**Pull request by proggeramlug**
Fixes #6073.

## Problem

BigInt is a fixed 16-limb / 1024-bit two's-complement value (`BigIntHeader { limbs: [u64; 16] }`). Arithmetic past ±2^1023 silently wrapped like an `i1024` instead of being arbitrary precision, corrupting results with no throw:

```js
(2n ** 1024n) > 1e308   // node: true   perry (before): false  (2^1024 wrapped the sign bit)
```

Any product/power/sum exceeding the range (e.g. `2n ** 1024n`) returned a wrong answer silently.

## Approach

Full arbitrary precision is an XL, multi-file initiative (variable-length limbs + length-aware rewrites of every arith/bit/compare/parse routine, codegen boundary types, GC variable-size alloc, cross-thread serialization, typed-array interop). This PR is the **self-contained stopgap** called out in the issue: detect the overflow that was being silently dropped and throw `RangeError: "Maximum BigInt size exceeded"` (V8's own wording for exceeding its precision cap) instead of wrapping. This converts "silently wrong answer" into a loud, catchable failure with no NaN-box/GC/codegen changes — entirely within `bigint.rs`.

The threshold (±2^1023) is lower than V8's, but the observable shape (a `RangeError` on overflow) matches. In-range values — including secp256k1 intermediate products (~512-bit) — are unaffected.

## What throws now

A shared `magnitude_fits_1024(mag, negative)` predicate encodes the exact representable range `[-2^1023, 2^1023 - 1]` (including the lone `-2^1023` endpoint). Every overflow-capable entry point checks it:

- **`js_bigint_add` / `js_bigint_sub`** — textbook signed-overflow sign rule.
- **`js_bigint_mul` + `js_bigint_pow`** — shared `mul_limbs_checked` retains the full 2N-limb product and checks it fits. `pow`'s square-and-multiply loop is reworked to skip the final unused base-squaring, so an in-range result like `(2^300)^3` doesn't false-throw on the never-used `(2^300)^4`.
- **`js_bigint_shl`** (and `>>` with a negative count) — `shl_checked`.
- **`js_bigint_neg`** — catches negating `-2^1023` (the sole value whose two's-complement negation is itself).
- **String parsing** — `parse_radix_digits` (the `123n` / `BigInt("…")` path) and BN.js-compat `js_bigint_from_string_radix` (hex + general radix).
- **`BigInt(number)`** — `limbs_from_integer_f64` (e.g. `BigInt(1e308)`).

Operations that cannot overflow the range are untouched: `div`, `mod`, `>>`, `&`, `|`, `^`, `~` (a bijection on the range), and the i64/i128 fast paths. `asUintN`/`asIntN` truncation is correct spec behavior and is left as-is.

## Testing

- 5 new `bigint.rs` unit tests: `magnitude_fits_1024` at every boundary (`<2^1023`, `==2^1023` per sign, `(2^1023,2^1024)`, `>=2^1024`), plus large-but-in-range `pow`/`mul`/`add`/`sub`/`shl` that must **not** false-throw (e.g. `2^1000`, `(2^500)^2`, `(2^300)^3`). Full `bigint` suite green.
- The throwing path can't be unit-tested (`js_throw` with no active try-frame calls `std::process::exit`), so it's verified end-to-end by compiling a probe: every overflow site throws `RangeError: Maximum BigInt size exceeded`, and every in-range case matches Node byte-for-byte (including the `-2^1023` endpoint and a secp256k1-scale 512-bit product).

No new gap test is added: the stopgap intentionally diverges from Node (which is arbitrary-precision) past 2^1023, so a Node-oracle gap test would diverge by design. The existing bigint gap fixtures all stay well within range and continue to pass.

No version bump / changelog per contributor-PR convention.

## Summary by CodeRabbit

* **Bug Fixes**
  * BigInt operations now strictly enforce the supported signed 1024-bit range.
  * Conversions, radix parsing, add/sub, negation, multiplication, exponentiation, and left/right shifts now throw `RangeError: Maximum BigInt size exceeded` on overflow instead of wrapping/truncating.
  * Improved handling for edge cases like negating the minimum representable value and detecting overflow in hex/other radices.
* **Tests**
  * Added coverage for the 1024-bit boundary conditions and end-to-end overflow scenarios (including pow/mul/add-sub/shl).

### CODE REVIEW

**Comment by proggeramlug** [comment]
@coderabbitai review

File: crates/perry-runtime/src/bigint.rs [hunk 1]
```diff
@@ -490,6 +542,10 @@ pub extern "C" fn js_bigint_from_string_radix(
                 }
                 *limb = value;
             }
+            // Any hex digits beyond the low 1024 bits overflow (#6073).
+            if chars.any(|c| c.is_ascii_hexdigit()) {
+                throw_bigint_overflow();
+            }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/bigint.rs]: Good catch — fixed in aaf1e116. The hex leftover check now throws only on a nonzero high-order digit (`matches!(c, '1'..='9' | 'a'..='f' | 'A'..='F')`), so excess leading zeros are accepted; the `magnitude_fits_1024` check still covers the [2^1023, 2^1024) range. Added a regression test and verified end-to-end (`BigInt("0x" + "0".repeat(300) + "beef")` → 48879).
